src/bookshelf/src/bksw_sec.erl

```diff
@@ -150,26 +179,89 @@ encode_sign_error_response(AccessKeyId, Signature,
     Req2 = wrq:set_resp_body(Body, Req1),
     {{halt, 403}, Req2, Context}.
 
-encode_access_denied_error_response(RequestId, Req0, Context) ->
-    Req1 = bksw_req:with_amz_id_2(Req0),
-    Body = bksw_xml:access_denied_error(RequestId),
-    Req2 = wrq:set_resp_body(Body, Req1),
-    {{halt, 403}, Req2, Context}.
+% split  "<bucketname>/<key>" (possibly leading and/or trailing /) into {"bucketname", "key"}
+% Path = "<bucketname>/<key>"
+-spec get_bucket_key(Path::string()) -> {string(), string()}.
+get_bucket_key(Path) ->
+    case string:lexemes(Path, "/") of
+        [            ] -> {"",     ""};
+        [Bucket      ] -> {Bucket, ""};
+        [Bucket | Key] -> {Bucket, filename:join(Key)}
+    end.
 
-split_authorization([$A, $W, $S, $\s, $: | Rest]) ->
-    [<<>>, Rest];
-split_authorization([$A, $W, $S, $\s  | Rest]) ->
-    string:tokens(Rest, ":").
+% get host and toggle the port (add port or remove it)
+-spec get_host_toggleport(string(), aws_config()) -> string().
+get_host_toggleport(Host, Config) ->
+    case string:split(Host, ":", trailing) of
+        [Host      ] ->
+            Port = integer_to_list(Config#aws_config.s3_port),
+            string:join([Host, Port], ":");
+        ["http",  _] ->
+            Port = integer_to_list(Config#aws_config.s3_port),
+            string:join([Host, Port], ":");
+        ["https", _] ->
+            Port = integer_to_list(Config#aws_config.s3_port),
+            string:join([Host, Port], ":");
+        [H,       _] ->
+            H
+    end.
 
-is_amz([$X, $-, $A, $m, $z, $-, $A, $c, $l | _]) ->
-    true;
-is_amz([$x, $-, $a, $m, $z, $-, $a, $c, $l | _]) ->
-    true;
-is_amz(_) ->
-    false.
+-spec host(tuple()) -> list().
+host(Req0) ->
+    wrq:get_req_header("Host", Req0).
+
+% most ways of getting the date/time seem problematic.  for instance, docs for
+% calendar:universal_time() and erlang:universaltime() say: 'Returns local time
+% if universal time is unavailable.'
+% since it is unknown which time would be used, we could use local time and
+% convert to universal.  however, local time could be an 'illegal' time wrt
+% universal time if switching to/from daylight savings time.
+%
+% note that we are neither measuring an elapsed time, nor determining an order
+% of events, nor creating a unique name, so time correction should not be an
+% issue.
+% https://erlang.org/doc/apps/erts/time_correction.html
+% 1 =< ExpiresSec =< 604800
+-spec is_expired(DateTimeString::string(), ExpiresSec::integer()) -> boolean().
+is_expired(DateTimeString, ExpiresSec) ->
+    [Y1, Y2, Y3, Y4, M1, M2, D1, D2, _, H1, H2, N1, N2, S1, S2, _] = DateTimeString,
```

Comment:
It does seem like we could use `ec_date:parse(Date::string()) -> datetime()` to get the datetime() and pass into `datetime_to_gregorian_seconds`?

Reply:
Doesn't look like `ec_date:parse` parses the same format of string:
```
13> c(ec_date).
{ok,ec_date}
14> ec_date:parse("20150830T123600Z"). 
** exception throw: {ec_date,{bad_date,"20150830T123600Z"}}
     in function  ec_date:do_parse/3 (ec_date.erl, line 90

18> bksw_sec:is_expired("20150830T123600Z", 0). 
true
```

Reply:
I added a note documenting the string format used.